specs/dwritecore/DWriteCore.md

```diff
@@ -1050,9 +1230,185 @@ int __cdecl wmain(int argc, WCHAR** argv)
+    deviceContext->GetTransform(&transform);
+
+    // Multiply the world transform by the DPI scale to get a world and DPI transform.
+    float dpiX, dpiY;
+    deviceContext->GetDpi(&dpiX, &dpiY);
+    transform = transform * D2D1::Matrix3x2F::Scale(dpiX / 96, dpiY / 96);
+
+    // Combination of image formats we support.
+    constexpr DWRITE_GLYPH_IMAGE_FORMATS desiredGlyphImageFormats =
+        DWRITE_GLYPH_IMAGE_FORMATS_TRUETYPE |
+        DWRITE_GLYPH_IMAGE_FORMATS_CFF |
+        DWRITE_GLYPH_IMAGE_FORMATS_COLR |
+        DWRITE_GLYPH_IMAGE_FORMATS_SVG |
+        DWRITE_GLYPH_IMAGE_FORMATS_PNG |
+        DWRITE_GLYPH_IMAGE_FORMATS_JPEG |
+        DWRITE_GLYPH_IMAGE_FORMATS_TIFF |
+        DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8 |
+        DWRITE_GLYPH_IMAGE_FORMATS_COLR_PAINT_TREE;
+
+    // Color glyph runs of type DWRITE_GLYPH_IMAGE_FORMATS_COLR_PAINT_TREE will be
+    // rendered using D2D's DrawPaintGlyphRun, so get the paint feature level from D2D.
+    DWRITE_PAINT_FEATURE_LEVEL paintFeatureLevel = deviceContext->GetColorPaintFeatureLevel();
+
+    // Call TranslateColorGlyphRun to translate the monochrome "base" glyph run
+    // into a sequence of color glyph runs.
+    wil::com_ptr<IDWriteColorGlyphRunEnumerator1> colorRunEnumerator;
+
+    HRESULT hr = dwriteFactory->TranslateColorGlyphRun(
+        baselineOrigin,
+        &glyphRun,
+        /*glyphRunDescription*/ nullptr,
+        desiredGlyphImageFormats,
+        paintFeatureLevel,
+        DWRITE_MEASURING_MODE_NATURAL,
+        reinterpret_cast<DWRITE_MATRIX const*>(&transform),
+        colorPaletteIndex,
+        &colorRunEnumerator
+    );
+
+    // Handle DWRITE_E_NOCOLOR, which is returned if the base glyph run does not
+    // have any color glyphs in the desired glyph image formats. In this case,
+    // just render the base glyph run as a monochrome glyph run.
+    if (hr == DWRITE_E_NOCOLOR)
+    {
+        deviceContext->DrawGlyphRun(
+            baselineOrigin,
+            &glyphRun,
+            foregroundBrush,
+            DWRITE_MEASURING_MODE_NATURAL
+        );
+        return S_OK;
+    }
+
+    // Any other error returned by TranslateColorGyphRun is a failure.
+    RETURN_IF_FAILED(hr);
+
+    // Solid color brush to be lazily created if needed.
+    wil::com_ptr<ID2D1SolidColorBrush> solidBrush;
+
+    // Use the returned enumerator object to iterate over the color glyph runs.
+    for (;;)
+    {
+        // Advanced to the first or next run.
+        BOOL haveRun;
+        RETURN_IF_FAILED(colorRunEnumerator->MoveNext(&haveRun));
+        if (!haveRun)
+            break;
+
+        // Get a pointer to the color glyph run structure.
+        DWRITE_COLOR_GLYPH_RUN1 const* colorGlyphRun;
+        RETURN_IF_FAILED(colorRunEnumerator->GetCurrentRun(&colorGlyphRun));
+
+        // Determine the brush to use for this color glyph run.
+        ID2D1Brush* runBrush;
+        if (colorGlyphRun->paletteIndex == 0xFFFF)
```

Comment:
Where did this constant come from? Was it somehow defined by the caller in the original glyph run, or is it some systemwide API magic constant?

Reply:
It's a magic palette index meaning the foreground color.

Reply:
This feels like it warrants a constexpr definition, or at least, a #define.

Reply:
This is in the existing API, but I agree a #define would make sense.
Maybe DWRITE_PALETTE_INDEX_FOREGROUND?

Reply:
I went with `DWRITE_NO_PALETTE_INDEX` since it's more general. It just means a palette index is not specify, which depending on context could mean use the foreground color or just that the color is not relevant.